common/http/platform/beast/http_resumer.cpp

```diff
@@ -0,0 +1,454 @@
+	if (!exp_content_range) {
+		client_.logger_.Error(exp_content_range.error().String());
+		client_.Cancel();
+		client_.state_.err = exp_content_range.error();
+		return;
+	}
+
+	auto content_range = exp_content_range.value();
+	if (content_range.size != 0 && content_range.size != client_.state_.content_length) {
+		auto err = MakeError(
+			DownloadResumerError,
+			"Size of artifact changed after download was resumed (expected "
+				+ to_string(client_.state_.content_length) + ", got "
+				+ to_string(content_range.size) + ")");
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		client_.state_.err = err;
+		return;
+	}
+	if ((content_range.range_end != client_.state_.content_length - 1)
+		|| (content_range.range_start > client_.state_.offset)) {
+		auto err = MakeError(
+			DownloadResumerError,
+			"HTTP server did not return expected range. Expected "
+				+ to_string(client_.state_.offset) + "-"
+				+ to_string(client_.state_.content_length - 1) + ", got "
+				+ to_string(content_range.range_start) + "-" + to_string(content_range.range_end));
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		client_.state_.err = err;
+	} else if (content_range.range_start < client_.state_.offset) {
+		// The server resumed from an earlier offset, but we can still use the data
+		auto buffer = client_.state_.buffer.get();
+		auto overriden = client_.state_.offset - content_range.range_start;
+		client_.logger_.Warning(
+			"HTTP server returned a range with an earlier start. Expected "
+			+ to_string(client_.state_.offset) + ", got " + to_string(content_range.range_start)
+			+ ". The last " + to_string(overriden) + " bytes will be overriden");
+		buffer->erase(buffer->end() - overriden, buffer->end());
+		client_.state_.offset = buffer->size();
+	}
+}
+
+events::EventHandler BodyHandlerFunctor::GetWaitCallback() {
+	HeaderHandlerFunctor resumer_next_header_handler {
+		user_header_handler_, user_body_handler_, user_request_, client_};
+	BodyHandlerFunctor resumer_next_body_handler {
+		user_header_handler_, user_body_handler_, user_request_, client_};
+
+	return [resumer_next_header_handler,
+			resumer_next_body_handler,
+			range_req = RemainingRangeRequest()](error::Error err) {
+		if (err != error::NoError) {
+			auto err_user =
+				MakeError(DownloadResumerError, "Unexpected error in wait timer: " + err.String());
+			resumer_next_body_handler.client_.logger_.Error(err_user.String());
+			resumer_next_body_handler.client_.Cancel();
+			resumer_next_body_handler.user_body_handler_(expected::unexpected(err_user));
+			return;
+		}
+
+		auto next_call_err = resumer_next_body_handler.client_.client_.AsyncCall(
+			range_req, resumer_next_header_handler, resumer_next_body_handler);
+		if (next_call_err != error::NoError) {
+			auto err_user = MakeError(
+				DownloadResumerError,
+				"Failed to copy data to schedule the next resumer call: " + next_call_err.String());
```

Comment:
Not sure if I understand this error message. What does "copy data" have to do with a failure from `AsyncCall`?

Reply:
Bad copy paste, it is meant to say `Failed to schedule the next resumer call`